frame/nfts/src/features/atomic_swap.rs

```diff
@@ -0,0 +1,152 @@
+
+		let now = frame_system::Pallet::<T>::block_number();
+		let deadline = maybe_duration.map(|d| d.saturating_add(now));
+
+		PendingSwapOf::<T, I>::insert(
+			&collection_id,
+			&item_id,
+			PendingSwap {
+				desired_collection: desired_collection_id,
+				desired_item: desired_item_id,
+				price: maybe_price,
+				deadline,
+			},
+		);
+
+		Self::deposit_event(Event::SwapCreated {
+			collection: collection_id,
+			item: item_id,
+			desired_collection: desired_collection_id,
+			desired_item: desired_item_id,
+			price: maybe_price,
+			deadline,
+		});
+
+		Ok(())
+	}
+
+	pub fn do_cancel_swap(
+		sender: T::AccountId,
+		collection_id: T::CollectionId,
+		item_id: T::ItemId,
+	) -> DispatchResult {
+		let item = Item::<T, I>::get(&collection_id, &item_id).ok_or(Error::<T, I>::UnknownItem)?;
+		let swap = PendingSwapOf::<T, I>::get(&collection_id, &item_id)
+			.ok_or(Error::<T, I>::UnknownSwap)?;
+
+		let is_past_deadline = if let Some(deadline) = swap.deadline {
+			let now = frame_system::Pallet::<T>::block_number();
+			now > deadline
+		} else {
+			false
+		};
+
+		if !is_past_deadline {
+			ensure!(item.owner == sender, Error::<T, I>::NoPermission);
+		}
+
+		PendingSwapOf::<T, I>::remove(&collection_id, &item_id);
+
+		Self::deposit_event(Event::SwapCancelled {
+			collection: collection_id,
+			item: item_id,
+			desired_collection: swap.desired_collection,
+			desired_item: swap.desired_item,
+			price: swap.price,
+			deadline: swap.deadline,
+		});
+
+		Ok(())
+	}
+
+	pub fn do_claim_swap(
+		sender: T::AccountId,
+		send_collection_id: T::CollectionId,
+		send_item_id: T::ItemId,
+		receive_collection_id: T::CollectionId,
+		receive_item_id: T::ItemId,
+		maybe_receive_amount: Option<ItemPrice<T, I>>,
+	) -> DispatchResult {
+		let send_item = Item::<T, I>::get(&send_collection_id, &send_item_id)
+			.ok_or(Error::<T, I>::UnknownItem)?;
+		let receive_item = Item::<T, I>::get(&receive_collection_id, &receive_item_id)
+			.ok_or(Error::<T, I>::UnknownItem)?;
+		let swap = PendingSwapOf::<T, I>::get(&receive_collection_id, &receive_item_id)
+			.ok_or(Error::<T, I>::UnknownSwap)?;
+
+		ensure!(send_item.owner == sender, Error::<T, I>::NoPermission);
+		ensure!(swap.desired_collection == send_collection_id, Error::<T, I>::UnknownSwap);
+		ensure!(swap.desired_item == send_item_id, Error::<T, I>::UnknownSwap);
+		ensure!(swap.price == maybe_receive_amount, Error::<T, I>::UnknownSwap);
+
+		if let Some(deadline) = swap.deadline {
+			let now = frame_system::Pallet::<T>::block_number();
+			ensure!(deadline >= now, Error::<T, I>::DeadlineExpired);
+		}
+
+		if let Some(amount) = swap.price {
+			T::Currency::transfer(&receive_item.owner, &send_item.owner, amount, KeepAlive)?;
+		}
+
+		Self::do_transfer(send_collection_id, send_item_id, receive_item.owner.clone(), |_, _| {
+			Ok(())
+		})?;
+		Self::do_transfer(
+			receive_collection_id,
+			receive_item_id,
+			send_item.owner.clone(),
+			|_, _| Ok(()),
+		)?;
```

Comment:
what happens if the second transfer fails? does the transaction kick in and roll back the first transfer?

Reply:
I guess I'm saying that this would be a great place to add a comment stating why this is atomic.

Reply:
It's the default behaviour across all the pallets now: if some action fails, the whole state gets rolled back

Reply:
so isn't everything atomic then? ^_^
I guess people name it as such...